File: app/src/main/java/com/example/note/domain/FirestoreNotesRepository.java
package com.example.note.domain;

import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;

public class FirestoreNotesRepository implements NotesRepository {
    private final static String NOTES = "notes";
    private final static String TITLE = "title";
    private final static String DESCRIPTION = "description";
    private final static String DATE = "date";

    private final FirebaseFirestore firestore = FirebaseFirestore.getInstance();

    @Override
    public void getNotes(Callback<List<Note>> callback) {
        firestore.collection(NOTES).get().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                final List<Note> temp = new ArrayList<>();
                List<DocumentSnapshot> documentSnapshots = task.getResult().getDocuments();

                for (DocumentSnapshot documentSnapshot : documentSnapshots) {
                    String id = documentSnapshot.getId();
                    String title = documentSnapshot.getString(TITLE);
                    String description = documentSnapshot.getString(DESCRIPTION);
                    Date date = documentSnapshot.getDate(DATE);

                    temp.add(new Note(id, title, description, date));
                }

                callback.onSuccess(temp);
            } else {
                callback.onError(task.getException());
            }
        });
    }

    @Override
    public void add(String title, String description, Callback<Note> callback) {
        HashMap<String, Object> data = new HashMap<>();
        Date date = new Date();

        data.put(TITLE, title);
        data.put(DESCRIPTION, description);
        data.put(DATE, date);

        firestore.collection(NOTES).add(data).addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                callback.onSuccess(new Note(task.getResult().getId(), title, description, date));
            } else {
                callback.onError(task.getException());
            }
        });
    }

    @Override
    public void delete(Note item, Callback<Object> callback) {
        firestore.collection(NOTES).document(item.getId()).delete().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                callback.onSuccess(new Object());
            } else {
                callback.onError(task.getException());
            }
        });
    }
}

File: app/src/main/java/com/example/note/domain/FirestoreNotesRepository.kt
package com.example.note.domain

import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

class FirestoreNotesRepository : NotesRepository {
    companion object {
        private const val NOTES = "notes"
        private const val TITLE = "title"
        private const val DESCRIPTION = "description"
        private const val DATE = "date"
    }

    private val firestore = FirebaseFirestore.getInstance()

    override fun getNotes(callback: Callback<List<Note>>) {
        firestore.collection(NOTES).get().addOnCompleteListener {
            if (it.isSuccessful) {
                val temp = ArrayList<Note>()
                val documentSnapshots: List<DocumentSnapshot> = it.result!!.documents

                for (documentSnapshot in documentSnapshots) {
                    val id = documentSnapshot.id
                    val title = documentSnapshot.getString(TITLE)
                    val description = documentSnapshot.getString(DESCRIPTION)
                    val date = documentSnapshot.getDate(DATE)

                    //TODO(check title, description, date for nullable)
                    temp.add(Note(id, title!!, description!!, date!!))
                }

                callback.onSuccess(temp)
            } else {
                it.exception?.let { exception -> callback.onError(exception) }
            }
        }
    }

    override fun add(title: String, description: String, callback: Callback<Note>) {
        val data = HashMap<String, Any>()
        val date = Date()

        data[TITLE] = title
        data[DESCRIPTION] = description
        data[DATE] = date

        firestore.collection(NOTES).add(data).addOnCompleteListener {
            if (it.isSuccessful) {
                callback.onSuccess(Note(it.result!!.id, title, description, date))
            } else {
                it.exception?.let { it1 -> callback.onError(it1) }
            }
        }
    }

    override fun delete(item: Note, callback: Callback<Any>) {
        firestore.collection(NOTES).document(item.id).delete().addOnCompleteListener {
            if (it.isSuccessful) {
                callback.onSuccess(Any())
            } else {
                it.exception?.let { it1 -> callback.onError(it1) }
            }
        }
    }
}